### PR TITLE
Dragonball: support resize memory

### PR DESCRIPTION
**Pull request by HerlinCoder**
we introduce two virtio devices to support resizing memory:
1. virtio-balloon device. virtio-balloon device could reclaim memory from guest to host.
2. virtio-mem device.  virtio-mem device could hot-plug more memory blocks to guest and could also hot-unplug them from guest.

These patches support ability of Dragonball resizing memory, and expose interface to runtime-rs.

### CODE REVIEW

File: src/dragonball/src/api/v1/vmm_action.rs [hunk 1]
```diff
@@ -35,9 +35,14 @@ pub use crate::device_manager::virtio_net_dev_mgr::{
 #[cfg(feature = "virtio-vsock")]
 pub use crate::device_manager::vsock_dev_mgr::{VsockDeviceConfigInfo, VsockDeviceError};
```
Comment (ZizhengBian) [inline src/dragonball/src/api/v1/vmm_action.rs:37]: no blank line here?

File: src/dragonball/src/api/v1/vmm_action.rs [hunk 2]
```diff
@@ -1456,4 +1546,44 @@ mod tests {
             t.check_request();
         }
     }
+
+    #[cfg(feature = "virtio-balloon")]
+    #[test]
```
Comment (ZizhengBian) [inline src/dragonball/src/api/v1/vmm_action.rs:1551]: add test cases for virtio-mem

File: src/dragonball/src/device_manager/balloon_dev_mgr.rs [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+
+use crate::address_space_manager::GuestAddressSpaceImpl;
+use crate::config_manager::{ConfigItem, DeviceConfigInfo, DeviceConfigInfos};
+use crate::device_manager::DbsMmioV2Device;
+use crate::device_manager::{DeviceManager, DeviceMgrError, DeviceOpContext};
+
+// The flag of whether to use the shared irq.
+const USE_SHARED_IRQ: bool = true;
+// The flag of whether to use the generic irq.
+const USE_GENERIC_IRQ: bool = false;
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/balloon_dev_mgr.rs:17]: Does kata support generic irq now?

File: src/dragonball/src/device_manager/balloon_dev_mgr.rs [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+    /// The balloon device was already used.
+    #[error("the virtio-balloon sock path was already added to a different device")]
+    BalloonDeviceAlreadyExists,
+
+    /// Cannot perform the requested operation after booting the microVM.
+    #[error("the update operation is not allowed after boot")]
+    UpdateNotAllowedPostBoot,
+
+    /// guest memory error
+    #[error("failed to access guest memory")]
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/balloon_dev_mgr.rs:31]: print error details here, same below

File: src/dragonball/src/device_manager/balloon_dev_mgr.rs [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+                if let Some(mmio_dev) = device.as_any().downcast_ref::<DbsMmioV2Device>() {
+                    let guard = mmio_dev.state();
+                    let inner_dev = guard.get_inner_device();
+                    if let Some(balloon_dev) = inner_dev
+                        .as_any()
+                        .downcast_ref::<Balloon<GuestAddressSpaceImpl>>()
+                    {
+                        return balloon_dev
+                            .set_size(config.size_mib)
+                            .map_err(|_e| BalloonDeviceError::ResizeFailed);
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/balloon_dev_mgr.rs:255]: Pls return error message to caller

File: src/dragonball/src/device_manager/balloon_dev_mgr.rs [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+                    }
+                }
+                Ok(())
+            }
+            None => Err(BalloonDeviceError::InvalidDeviceId(
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/balloon_dev_mgr.rs:260]: if let is better?

File: src/dragonball/src/device_manager/mem_dev_mgr.rs [hunk 7]
```diff
@@ -0,0 +1,742 @@
...
+use crate::vm::VmConfigInfo;
+
+// The flag of whether to use the shared irq.
+const USE_SHARED_IRQ: bool = true;
+// The flag of whether to use the generic irq.
+const USE_GENERIC_IRQ: bool = false;
+
+/// Madvise flag for nocma memory allocation
+#[cfg(target_arch = "aarch64")]
+const MADV_NOCMA: i32 = 25;
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mem_dev_mgr.rs:36]: does we really support nocma mode in kata?

File: src/dragonball/src/device_manager/mem_dev_mgr.rs [hunk 8]
```diff
@@ -0,0 +1,742 @@
...
+
+const HUGE_PAGE_2M: usize = 0x200000;
+
+// max numa node ids on host
+const MAX_NODE: u32 = 64;
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mem_dev_mgr.rs:41]: MAX_NODE_ID or MAX_NODE_NR?

File: src/dragonball/src/device_manager/mem_dev_mgr.rs [hunk 9]
```diff
@@ -0,0 +1,742 @@
...
+        if !cfg!(feature = "hotplug") && ctx.is_hotplug {
+            error!(ctx.logger(), "hotplug feature has been disabled.";
+            "subsystem" => "virito-mem");
+            return Err(MemDeviceError::UpdateNotAllowedPostBoot);
+        }
+
+        let epoll_mgr = ctx.epoll_mgr.clone().ok_or(MemDeviceError::ResizeFailed)?;
+
+        // If the id of the drive already exists in the list, the operation is update.
+        match self.get_index_of_mem_dev(&mem_cfg.mem_id) {
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mem_dev_mgr.rs:173]: if let is better?

File: src/dragonball/src/device_manager/mem_dev_mgr.rs [hunk 10]
```diff
@@ -0,0 +1,742 @@
...
+                if let Some(mmio_dev) = device.as_any().downcast_ref::<DbsMmioV2Device>() {
+                    let guard = mmio_dev.state();
+                    let inner_dev = guard.get_inner_device();
+                    if let Some(mem_dev) = inner_dev
+                        .as_any()
+                        .downcast_ref::<Mem<GuestAddressSpaceImpl>>()
+                    {
+                        return mem_dev
+                            .set_requested_size(config.size_mib)
+                            .map_err(|_e| MemDeviceError::ResizeFailed);
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mem_dev_mgr.rs:346]: Pls return error info to callers

File: src/dragonball/src/device_manager/mem_dev_mgr.rs [hunk 11]
```diff
@@ -0,0 +1,742 @@
...
+
+        Ok(())
+    }
+
+    fn update_memory_size(
+        &self,
+        config: &MemDeviceConfigInfo,
+    ) -> std::result::Result<(), MemDeviceError> {
+        match self.get_index_of_mem_dev(&config.mem_id) {
+            Some(index) => {
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mem_dev_mgr.rs:332]: if let is better?

File: src/dragonball/src/device_manager/mem_dev_mgr.rs [hunk 12]
```diff
@@ -0,0 +1,742 @@
...
+    /// Create a new `MemDeviceMgr` object..
+    fn default() -> Self {
+        MemDeviceMgr {
+            info_list: DeviceConfigInfos::new(),
+            use_shared_irq: USE_SHARED_IRQ,
+        }
+    }
+}
+
+pub(crate) struct MemoryRegionFactory {
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mem_dev_mgr.rs:366]: Do we need so many pub(crate)?

File: src/dragonball/src/device_manager/mem_dev_mgr.rs [hunk 13]
```diff
@@ -0,0 +1,742 @@
...
+            kvm_slot,
+            0,
+            guest_addr.raw_value(),
+            region_len,
+            host_addr;
+            "subsystem" => "virito-mem"
+        );
+
+        // All value should be valid.
+        let memory_region = Arc::new(GuestRegionMmap::new(mmap_region, guest_addr).unwrap());
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mem_dev_mgr.rs:574]: is it safe to unwrap?

File: src/dragonball/src/device_manager/mod.rs [hunk 14]
```diff
@@ -874,6 +942,24 @@ impl DeviceManager {
...
 
+    /// Create an Virtio MMIO transport layer device for the virtio backend device with configure
+    /// change notification enabled.
+    pub fn create_mmio_virtio_device_with_device_change_notification(
+        device: DbsVirtioDevice,
+        ctx: &mut DeviceOpContext,
+        use_shared_irq: bool,
+        use_generic_irq: bool,
+    ) -> std::result::Result<Arc<DbsMmioV2Device>, DeviceMgrError> {
+        let features = DRAGONBALL_FEATURE_PER_QUEUE_NOTIFY;
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mod.rs:953]: Does kata really support this feature?

File: src/dragonball/src/device_manager/balloon_dev_mgr.rs (hunk 3 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/balloon_dev_mgr.rs:17]: we will refactor this in later patches

File: src/dragonball/src/device_manager/mem_dev_mgr.rs (hunk 7 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/mem_dev_mgr.rs]: removed

File: src/dragonball/src/device_manager/mem_dev_mgr.rs (hunk 8 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/mem_dev_mgr.rs:37]: same as address space manager. I don't feel misleading here

File: src/dragonball/src/device_manager/mem_dev_mgr.rs (hunk 9 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/mem_dev_mgr.rs]: fixed

File: src/dragonball/src/api/v1/vmm_action.rs (hunk 1 above)
Comment (HerlinCoder) [reply src/dragonball/src/api/v1/vmm_action.rs]: done

File: src/dragonball/src/api/v1/vmm_action.rs (hunk 2 above)
Comment (HerlinCoder) [reply src/dragonball/src/api/v1/vmm_action.rs:1589]: done

File: src/dragonball/src/device_manager/mem_dev_mgr.rs (hunk 10 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/mem_dev_mgr.rs]: done

File: src/dragonball/src/device_manager/mem_dev_mgr.rs (hunk 11 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/mem_dev_mgr.rs]: done

File: src/dragonball/src/device_manager/mem_dev_mgr.rs (hunk 13 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/mem_dev_mgr.rs]: fixed

File: src/dragonball/src/device_manager/mod.rs (hunk 14 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/mod.rs:953]: Other devices also has this feature.

**Comment by ZizhengBian** [comment]
LGTM！

**Review by ZizhengBian — APPROVED** [review]

**Review by studychao — COMMENTED** [review]
Thanks @HerlinCoder , a few comments.

**Review by studychao — COMMENTED** [review]
Thanks @HerlinCoder , a few comments.

File: src/dragonball/src/device_manager/balloon_dev_mgr.rs [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+// The flag of whether to use the shared irq.
+const USE_SHARED_IRQ: bool = true;
+// The flag of whether to use the generic irq.
+const USE_GENERIC_IRQ: bool = false;
+
+/// Errors associated with `BalloonDeviceConfig`.
+#[derive(Debug, thiserror::Error)]
+pub enum BalloonDeviceError {
+    /// The balloon device was already used.
+    #[error("the virtio-balloon sock path was already added to a different device")]
```
Comment (studychao) [inline src/dragonball/src/device_manager/balloon_dev_mgr.rs:24]: Why this is `sock path`?
I read through the code, mostly is `id` conflicts but I don't see any `sock path`

File: src/dragonball/src/api/v1/vmm_action.rs [hunk 16]
```diff
@@ -648,6 +680,62 @@ impl VmmService {
...
+
+    #[cfg(feature = "virtio-mem")]
+    fn add_mem_device(
+        &mut self,
+        vmm: &mut Vmm,
+        event_mgr: &mut EventManager,
+        config: MemDeviceConfigInfo,
+    ) -> VmmRequestResult {
+        let vm = vmm.get_vm_mut().ok_or(VmmActionError::InvalidVMID)?;
+
```
[the remark this answers is not in the document]
Comment (HerlinCoder) [reply src/dragonball/src/api/v1/vmm_action.rs:692]: mem device only operate memory belong to itself. we will support mem device prealloc in the future.

File: src/dragonball/src/device_manager/balloon_dev_mgr.rs [hunk 17]
```diff
@@ -0,0 +1,435 @@
...
+            "subsystem" => "balloon_dev_mgr",);
+            return Err(BalloonDeviceError::UpdateNotAllowedPostBoot);
+        }
+
+        let epoll_mgr = ctx
+            .get_epoll_mgr()
+            .map_err(BalloonDeviceError::DeviceManager)?;
+
+        // If the id of the drive already exists in the list, the operation is update.
+        if let Some(_index) = self.get_index_of_balloon_dev(&balloon_cfg.balloon_id) {
```
[the remark this answers is not in the document]
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/balloon_dev_mgr.rs]: done

File: src/dragonball/src/device_manager/mem_dev_mgr.rs [hunk 18]
```diff
@@ -0,0 +1,736 @@
...
+        if !cfg!(feature = "hotplug") && ctx.is_hotplug {
+            error!(ctx.logger(), "hotplug feature has been disabled.";
+            "subsystem" => "virito-mem");
+            return Err(MemDeviceError::UpdateNotAllowedPostBoot);
+        }
+
+        let epoll_mgr = ctx.get_epoll_mgr().map_err(MemDeviceError::DeviceManager)?;
+
+        // If the id of the drive already exists in the list, the operation is update.
+        if let Some(_index) = self.get_index_of_mem_dev(&mem_cfg.mem_id) {
```
[the remark this answers is not in the document]
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/mem_dev_mgr.rs]: refactor this function to make it usable

File: src/dragonball/src/device_manager/balloon_dev_mgr.rs (hunk 15 above)
Comment (HerlinCoder) [reply src/dragonball/src/device_manager/balloon_dev_mgr.rs]: you're right! this is fixed

**Review by studychao — APPROVED** [review]
LGTM

**Review by Tim-0731-Hzt — APPROVED** [review]

**Review by quanweiZhou — APPROVED** [review]

**Pull request merged**